WooCommerce: Update variation component to use product redux state

justinshreve opened the pull request:
This PR updates the variation types component to use the product attribute actions to update a product's attributes/variation types as the user fills out their information. Previously the component (my first) just had internal state, which has been removed in this PR & converted to a functional component.

To Test:

* `make run`
* Go to http://calypso.localhost:3000/store/products/add/
* Toggle the "has variations" toggle and make sure this works.
* Type in a variation type name like 'Color' and make sure the name field works.
* Add options like Red & Blue to make sure the options field works.
* Click "Add another variation" to make sure new rows can be added.

allendav commented on client/extensions/woocommerce/app/products/product-form-variations-card.js:
```diff
@@ -54,13 +50,16 @@ export default class ProductFormVariationCard extends Component {
+				{ 'variable' === product.type && (
+					<ProductVariationTypesForm
+						product={ product }
+						editProductAttribute={ this.props.editProductAttribute }
```
Why not add `editProductAttribute` to the destructuring above to keep the code here as brief as possible?

justinshreve replied:
Done.

allendav commented:
Toggling the variation section on the first time results in a JS error:

<img width="1171" alt="screen shot 2017-04-21 at 3 51 18 pm" src="https://cloud.githubusercontent.com/assets/1595739/25298530/8024179e-26aa-11e7-8093-d8d286ab47ef.png">

allendav commented:
The following error is thrown for each character typed in the Name box

```
index.js:36 Error: Array#find is not supported on all browsers. You must use a replacement method from lodash.
    at Array.eval (app:///./client/lib/wrap-es6-functions/index.js:35:13)
    at ProductVariationTypesForm.updateName (app:///./client/extensions/woocommerce/app/products/product-variation-types-form.js:111:61)
    at Object.executeOnChange (app:///./~/react-dom/lib/LinkedValueUtils.js:129:34)
    at ReactDOMComponent._handleChange (app:///./~/react-dom/lib/ReactDOMInput.js:222:38)
    at Object.ReactErrorUtils.invokeGuardedCallback (app:///./~/react-dom/lib/ReactErrorUtils.js:70:16)
    at executeDispatch (app:///./~/react-dom/lib/EventPluginUtils.js:85:21)
    at Object.executeDispatchesInOrder (app:///./~/react-dom/lib/EventPluginUtils.js:108:5)
    at executeDispatchesAndRelease (app:///./~/react-dom/lib/EventPluginHub.js:43:22)
    at executeDispatchesAndReleaseTopLevel (app:///./~/react-dom/lib/EventPluginHub.js:54:10)
    at Array.forEach (native)
```

coderkevin commented:
Rebased.

coderkevin commented:
>Toggling the variation section on the first time results in a JS error:
This should be fixed now as this reducer state will not be persisted anymore.

justinshreve commented:
Updates:

* Rebased
* Use lodash's find for updating variation type / attribute name.
* Fixed JS Warnings.
* Made ProductFormVariationsCard a Functional Component.

Should be ready for a look @allendav @coderkevin @DanReyLop

justinshreve commented:
@coderkevin The attributes are now cleared when setting the product type back to 'simple'.

coderkevin commented on client/extensions/woocommerce/app/products/product-form-variations-card.js:
```diff
@@ -11,60 +11,51 @@ import FoldableCard from 'components/foldable-card';
+		if ( 'variable' !== product.type ) {
+			editProduct( product, { type: 'variable' } );
+		} else {
+			editProduct( product, { type: 'simple', attributes: null } );
```
When we start to add non-variation attributes, this line will remove them. I don't think we have to solve this problem now, but I'd suggest adding a TODO comment above it to give yourself a reminder when you start implementing attributes.

justinshreve replied:
Good catch -- that can be next after variations is in, so I'll add the TODO comment and merge. Thanks!
